barretenberg/cpp/src/barretenberg/ecc/groups/affine_element_impl.hpp

```diff
@@ -218,22 +237,58 @@ affine_element<Fq, Fr, T> affine_element<Fq, Fr, T>::hash_to_curve(uint64_t seed
 
     bool y_bit = hash.get_bit(255);
 
-    Fq x_out = Fq(x_coordinate);
-    Fq y_out = (x_out.sqr() * x_out + T::b);
-    if constexpr (T::has_a) {
-        y_out += (x_out * T::a);
+    std::optional<affine_element> result = derive_from_x_coordinate(x_coordinate, y_bit);
+
+    if (result.has_value()) {
+        return result.value();
     }
+    return affine_element(0, 0);
+}
 
-    // When the sqrt of y_out doesn't exist, return 0.
-    auto [is_quadratic_remainder, y_out_] = y_out.sqrt();
-    if (!is_quadratic_remainder) {
-        return affine_element(Fq::zero(), Fq::zero());
+template <class Fq, class Fr, class T>
+template <typename>
+affine_element<Fq, Fr, T> affine_element<Fq, Fr, T>::hash_to_curve(const std::vector<uint8_t>& seed) noexcept
+{
+    std::vector<uint8_t> target_seed(seed);
+
+    // expand by 33 bytes to cover incremental hash attempts
+    const size_t seed_size = seed.size();
+    for (size_t i = 0; i < 33; ++i) {
+        target_seed.emplace_back(0);
     }
-    if (uint256_t(y_out_).get_bit(0) != y_bit) {
-        y_out_ = -y_out_;
+    uint16_t attempt_counter = 0;
+
+    while (true) {
+        auto hi = static_cast<uint8_t>(attempt_counter >> static_cast<uint16_t>(8));
+        auto lo = static_cast<uint8_t>(attempt_counter & static_cast<uint16_t>(0xff));
+        target_seed[seed_size] = hi;
+        target_seed[seed_size + 1] = lo;
+        target_seed[target_seed.size() - 1] = 0;
+        std::array<uint8_t, 32> hash_hi = sha256::sha256(target_seed);
+        target_seed[target_seed.size() - 1] = 1;
+        std::array<uint8_t, 32> hash_lo = sha256::sha256(target_seed);
+        std::vector<uint8_t> gg(hash_hi.begin(), hash_hi.end());
+        std::vector<uint8_t> ff(hash_lo.begin(), hash_lo.end());
+        uint256_t x_lo = 0;
+        uint256_t x_hi = 0;
+        // uint8_t* f = &hash_lo[0];
+        // uint8_t* g = &hash_hi[0];
+        read(ff, x_lo);
+        read(gg, x_hi);
+        // numeric::read(*f, x_lo);
+        // numeric::read(*g, x_hi);
+        uint512_t x_full(x_lo, x_hi);
+        Fq x(x_full);
+        bool sign_bit = false;
+        sign_bit = x_hi.get_bit(0);
```

Comment:
```suggestion
       // Extract high and low bytes from the attempt_counter
       auto uint8_t hi = static_cast<uint8_t>(attempt_counter >> 8);
       auto uint8_t lo = static_cast<uint8_t>(attempt_counter & 0xff);
        
        target_seed[seed_size] = hi;
        target_seed[seed_size + 1] = lo;
        
        // Create hashes whose input differs by only the last element
        target_seed.back() = 0;
        std::array<uint8_t, 32> hash_hi = sha256::sha256(target_seed);
        
        target_seed.back() = 1;
        std::array<uint8_t, 32> hash_lo = sha256::sha256(target_seed);
        
        std::vector<uint8_t> gg(hash_hi.begin(), hash_hi.end());
        std::vector<uint8_t> ff(hash_lo.begin(), hash_lo.end());
        
        uint256_t x_lo = 0;
        uint256_t x_hi = 0;
        
        read(ff, x_lo);
        read(gg, x_hi);
        
        uint512_t x_full(x_lo, x_hi);
        
        // Initialize Fq with a 512 bit integer by reducing it modulo the modulus `q`
        Fq x(x_full);
        
        bool sign_bit = false;
        sign_bit = x_hi.get_bit(0);
```

Reply:
Mainly added spacing for readability, not sure the static casts are needed, maybe its a clang error, and changed arr[size -1] to be arr.back() which returns a reference to the last element

Reply:
I refactored the method to remove the weird return(0, 0) thing at the end. lmk if you think it needs more changes